packages/firestore/test/unit/specs/perf_spec.test.ts

```diff
@@ -0,0 +1,216 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Query } from '../../../src/core/query';
+import { doc, orderBy, path } from '../../util/helpers';
+
+import { describeSpec, specTest } from './describe_spec';
+import { spec } from './spec_builder';
+
+const STEP_COUNT = 10;
+
+describeSpec('Performance Tests:', ['benchmark'], () => {
+  specTest('Insert a new document', [], () => {
+    let steps = spec().withGCEnabled(false);
+    for (let i = 0; i < STEP_COUNT; ++i) {
+      steps = steps.userSets(`collection/{i}`, { doc: i }).writeAcks(i);
+    }
+    return steps;
+  });
+
+  specTest('Insert a new document and wait for snapshot', [], () => {
+    let currentVersion = 1;
+    let steps = spec().withGCEnabled(false);
+
+    for (let i = 0; i < STEP_COUNT; ++i) {
+      const query = Query.atPath(path(`collection/${i}`));
+      const docLocal = doc(
+        `collection/${i}`,
+        0,
+        { doc: i },
+        { hasLocalMutations: true }
+      );
+      const docRemote = doc(`collection/${i}`, ++currentVersion, { doc: i });
+
+      steps = steps
+        .userListens(query)
+        .userSets(`collection/${i}`, { doc: i })
+        .expectEvents(query, {
+          added: [docLocal],
+          fromCache: true,
+          hasPendingWrites: true
+        })
+        .writeAcks(++currentVersion)
+        .watchAcksFull(query, ++currentVersion, docRemote)
+        .expectEvents(query, { metadata: [docRemote] })
+        .userUnlistens(query)
+        .watchRemoves(query);
+    }
+    return steps;
+  });
+
+  specTest('Watch has cached mutations', [], () => {
+    const cachedDocumentCount = 100;
+
+    const query = Query.atPath(path(`collection`)).addOrderBy(orderBy('v'));
+
+    let steps = spec().withGCEnabled(false);
+
+    const docs = [];
+
+    for (let i = 0; i < cachedDocumentCount; ++i) {
```

Comment:
It is a little tempting to somehow introduce the equivalent of for-loops into the spec test schema to avoid generating overly large / verbose JSON files with stuff like this and the STEP_COUNT stuff...  But that's probably a rabbit hole best avoided for now. :-)

Reply:
go/changestats

:)